File: leetcode/724-find-pivot-index/main.go
package main

import "fmt"

/*
	Questions to ask:
	- will there be more than one pivot? if yes, return which?
	- return what for an empty array?
	- return what for a single-element array?
	- will there be a list of the same number? e.g. [1,1,1,1,1] & [1,1,1,1,1,1]?

	i failed to come up with a correct approach to this question
	learned from others
	1.index: 0, num: 1, left: 0, right: 27
	2. index: 1, num: 7, left: 1, right: 20
	3. index: 2, num: 3, left: 8, right: 17
	4. index: 3, num: 6, left: 11, right: 11 <-- Found!!!
	Time	O(n)
	Space O(1)
*/
func pivotIndex(nums []int) int {
	right := 0
	for i := 0; i < len(nums); i++ {
		right += nums[i]
	}
	left := 0
	for i := 0; i < len(nums); i++ {
		right -= nums[i]
		if left == right {
			return i
		}
		left += nums[i]
	}
	return -1
}

func main() {
	fmt.Println(pivotIndex([]int{}))
	fmt.Println(pivotIndex([]int{1}))
	fmt.Println(pivotIndex([]int{1, 1}))
	fmt.Println(pivotIndex([]int{1, 2, 1}))
	fmt.Println(pivotIndex([]int{1, -1, 2, 2}))
	fmt.Println(pivotIndex([]int{1, -1, 2, 0, 4}))
	fmt.Println(pivotIndex([]int{1, 7, 3, 6, 5, 6}))
	fmt.Println(pivotIndex([]int{1, 7, 0, 0, 2, 6}))
	fmt.Println(pivotIndex([]int{1, 7, -1, 7, 0, 0, 2, 6}))
	fmt.Println(pivotIndex([]int{1, 1, 1, 1, 1}))
	fmt.Println(pivotIndex([]int{1, 1, 1, 1, 1, 1}))
	fmt.Println(pivotIndex([]int{-1, -1, -1, -1, -1}))
	fmt.Println(pivotIndex([]int{-1, -1, -1, -1, -1, -1}))
}

File: leetcode/724-find-pivot-index/main.py
"""
    1st: 2 arrays
    
    Time    O(3N)
    Space   O(2N)
    172 ms, faster than 56.15% 
"""


class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        n = len(nums)
        forward = n * [0]
        pfs = 0
        for i in range(n):
            pfs += nums[i]
            forward[i] = pfs
        backward = n * [0]
        sfs = 0
        for i in range(n-1, -1, -1):
            sfs += nums[i]
            backward[i] = sfs
        for i in range(n):
            if forward[i] == backward[i]:
                return i
        return -1


""" 
    2nd: 
    - similar to lc238, 724
    - at every index subtract prefix sum from the right until prefix sum == suffix sum

    [1, 7, 3, 6, 5, 6]
     1  8 11 17 22 28
             17 11 6
             ^
    
    Time    O(2N)
    Space   O(1)
    108 ms, faster than 98.64%
"""


class Solution(object):
    def pivotIndex(self, nums):
        pfs = sum(nums)
        sfs = 0
        for i in range(len(nums)):
            sfs += nums[i]
            if pfs == sfs:
                return i
            pfs -= nums[i]
        return -1
